implement retain release code motion.

trentxintong opened the pull request:
Implement a retain and release code motion pass.

Iterative data flow retain sinking and release hoisting.

This allows us to sink retains and hoist releases across harmless loops.  which is an improvement on the SILCodeMotion retain sinking and release hoisting.
    
This should eventually replace RR code motion in SILcodemotion and insertion point in ARCsequence opts.

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@shahmishal Any idea why this is very different from what i measured on my local imac machine. I do not see the improvements nor regressions reported in this commit

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@shahmishal I only see one RedundantLoadElimination.cpp.o compiled. Should not we compiled 2 versions of the compiler, i.e. one with the PR and one without and use those to compile the benchmarks.

https://ci.swift.org/job/swift-PR-osx-perf/30/

trentxintong commented:
@swift-ci Please ASAN test

trentxintong commented:
@swift-ci Please ASAN test

trentxintong commented:
@swift-ci Please ASAN test

trentxintong commented:
@swift-ci Please ASAN test

trentxintong commented:
@swift-ci Please ASAN test

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,997 @@
+/// \file
+///
+/// This pass moves retains down and releases up. This, hopefully, will help
+/// ASO to remove retain and release pairs without worrying too much about
```
Please expand out ASO here to make it more clear what you're talking about.

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please ASAN test

nadavrot commented on include/swift/SILOptimizer/Analysis/ARCAnalysis.h:
```diff
@@ -221,6 +221,18 @@ class ConsumedArgToEpilogueReleaseMatcher {
 
   bool hasBlock() const { return HasBlock; }
 
+  bool isEpilogueRelease(SILInstruction *I) const {
```
Can you please document this part?  Also, do you need the two checks?

rudkx replied:
Prior to looping through the instructions you should be able to get the exit block and see if I->getBlock() is the same block.For example you can refactor the code in ::recompute() to separate out the switch on ExitKind and selection of return/throw block, and then test against that.

trentxintong replied:
Or keep a SILBasicBlock *ThisBlock to represent the processed block by this ConsumedArgToEpilogueReleaseMatcher. And then we can remove the HasBlock variable.

nadavrot commented on include/swift/SILOptimizer/PassManager/Passes.def:
```diff
@@ -179,6 +179,12 @@ PASS(PredictableMemoryOptimizations, "predictable-memopt",
      "Predictable early memory optimizations")
 PASS(ReleaseDevirtualizer, "release-devirtualizer",
      "Devirtualize release-instructions")
+PASS(RetainSinking, "retain-sinking",
```
Why do you need three passes?  Can't you implement a single ARC code motion with three phases?

trentxintong replied:
We freeze epilogue releases (epilogue release matcher only finds the release in the return block currently) in till the late stage of the pipeline so we need a release hoisting and a late release hoisting. 

I experimented to hoist release first and run the first ARC pass and then sink the retains before running the second ARC pass in addSSAPasses gives the best performance.

rudkx replied:
I need to read through the rest of the passes, but I also wonder if this can be done in one pass instead of three.

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@shahmishal Thanks for fixing "please benchmark" this looks similar to what i saw on my local machine

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please smoke test

trentxintong commented:
@swift-ci Please smoke test

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please test linux

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please test linux

trentxintong commented:
@swift-ci Please ASAN test

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please ASAN test

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please test OSX

trentxintong commented:
@swift-ci Please test OS X

trentxintong commented:
@swift-ci Please test Linux

rudkx commented on include/swift/SILOptimizer/Analysis/ARCAnalysis.h:
```diff
@@ -221,6 +221,18 @@ class ConsumedArgToEpilogueReleaseMatcher {
+        continue;
+      if (X.second.size() > 1)
+        continue;
+      if (*X.second.begin() == I) 
```
Why isn't this != 1?

rudkx commented on include/swift/SILOptimizer/Analysis/ARCAnalysis.h:
```diff
@@ -221,6 +221,18 @@ class ConsumedArgToEpilogueReleaseMatcher {
+      if (X.second.size() == 0)
+        continue;
+      if (X.second.size() > 1)
+        continue;
```
Why not just check size() != 1?

trentxintong replied:
will do.

trentxintong commented:
@swift-ci Please test Linux

trentxintong commented:
@rudkx can you please review this pass ?

trentxintong commented:
@swift-ci Please test Linux

trentxintong commented:
@swift-ci Please test Linux

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please test OS X

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please benchmark

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1070 @@
+  // Otherwise create a release value.
+  return B.createReleaseValue(Loc, Ptr, Atomicity::Atomic);
+}
+
```
Rather than copy these from the ARCSequenceOpts, could you make a single shared versions of these?

It may even make sense to put these directly in SILBuilder if there were an unambiguous name that could be used (e.g. emitStrongRelease wouldn't be great because it sounds like it only creates strong_releases).

Perhaps emitReleaseValue() could work though, with the knowledge that it peepholes it into a simpler strong_release when appropriate?

trentxintong replied:
This can be done. In fact we have more places i think that makes uses of these functions in our code base. I will share them.

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1070 @@
+  /// NOTE: we could do the data flow with BBSetIn or BBSetOut, but that would
+  /// require us to create a temporary copy to check whether the BBSet has
+  /// changed after the genset and killset has been applied.
+  llvm::SmallBitVector BBSetIn;
```
Creating the temporary seems like the way to go here, otherwise you're creating and keeping N extra bit vectors around for an N-block function.

rudkx replied:
Have you measured the cardinality of any of these sets? For example getting a histogram of how large these tend to be at some point(s) in solving the equations, compared to the allocation size, could be interesting. I suspect these are relatively small so it might not matter, but in some cases sparse bit vectors can be a large win over dense ones both in terms of memory consumed and work involved in intersections and unions.

trentxintong replied:
I did. we rarely exceed 16 unique rc roots. usually, 1 - 3 rc roots. so we are allocating a lot of unused spaces here. but it does not take too much memory anyhow. and its allocation and deallocated through a BumpPtrAllocator which means allocations and deallocations are cheap most of the time .

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1070 @@
+
+  /// All the retains or releases originally in the function. Eventually
+  /// they will all be removed after all the new ones are generated.
+  llvm::DenseSet<SILInstruction *> RCInstructions;
```
You should consider SmallPtrSet instead of DenseSet. I believe DenseSet, like DenseMap, starts with size 64 which is going to be much larger than what you need in most cases.

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1070 @@
+  /// some RC instructions. If the basic block has neither, we do not need to
+  /// process the block again in the last iteration. We populate this set when
+  /// we compute the genset and killset.
+  llvm::DenseSet<SILBasicBlock *> InterestBlocks;
```
Same comment with respect to SmallPtrSet vs. DenseSet.

trentxintong commented:
@swift-ci Please benchmark

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1070 @@
+
+  /// Return the rc-identity root of the instruction.
+  SILValue getRCRoot(SILInstruction *I) {
+    return getRCRoot(I->getOperand(0));
```
This should be asserting on the kinds of instructions that are valid here. It's clear from the uses of the function that it's only called on certain instructions, but it should be made evident in the function body as well.

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1073 @@
+    return;
+
+  // Intersect in all predecessors' BBSetOuts.
+  State->BBSetIn.set();
```
You should initialize this with the first predecessor's out set to save from having to set all the bits and then immediately clear many of them.

rudkx commented on include/swift/SILOptimizer/Analysis/ARCAnalysis.h:
```diff
@@ -186,8 +186,12 @@ class ConsumedArgToEpilogueReleaseMatcher {
   bool HasBlock = false;
 
+  /// Eventually this will be used in place of HasBlock.
+  SILBasicBlock *ThisBlock;
```
You seem to already be using this. Is the comment out of date? It's preferred to not commit dead code (even commented), but instead introduce something when it's actually used. Is it possible to remove either ThisBlock or HasBlock and remove the other?

Also, a name more informative than ThisBlock would be helpful here.

trentxintong replied:
This comment is out of date. I will remove the HasBlock. actually i will be doing a refactoring on the retain release matcher sometime next week.

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please benchmark

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1084 @@
+    // Compute insertion point generated by the edge value transition.
+    // If there is a transition from 1 to 0, that means we have a partial
+    // merge, which means the retain can NOT be sunk to the current block.
+    // place it at the predecessors.
```
Nit: Comments should be complete sentences, appropriately capitalized. It's not clear if this was supposed to read "current block, so place it at the predecessors.", or if "place it at the predecessors" was meant to be a separate sentence. If the latter, please capitalize. Also, "in the predecessors" or "at the end of the predecessors" are both better options here.

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1084 @@
+    // Compute insertion point within the basic block. Process instructions in
+    // the basic block in reverse post-order fashion.
+    for (auto I = BB->begin(), E = BB->end(); I != E; ++I) {
+      for (unsigned i = 0; i < RCRootVault.size(); ++i) {
```
Rather than looping over every potentially set bit, you should use the find_first() and find_next() interface on SmallBitVector.

trentxintong replied:
Good point.

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1084 @@
+    }
+
+    // Lastly update the BBSetOut.
+    S->updateBBSetOut(S->BBSetIn);
```
Is this actually necessary?

trentxintong replied:
This is only necessary when we do the one-iteration data flow, i.e. no genset or killset generated and no convergence of the data flow.

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1084 @@
+        InsertPoints[RCRootVault[i]].push_back(Pred->getTerminator());
+      }
+    }
+
```
If BBSetIn is all-zero, do we actually need to continue processing (and walking the entire block) or can we just exit?

trentxintong replied:
If BBSetIn is all-ones (if thats what you mean), we can skip this loop, but we still need to execute the remaining of the function.

If BBSetIn is all-zeros, we need to execute this loop.

trentxintong replied:
I talked to @rudkx in person, he meant the loop on line 600. we need to walk the block because it may have retain instructions that "generate".

rudkx commented on lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp:
```diff
@@ -0,0 +1,1084 @@
+
+    // Compute insertion point within the basic block. Process instructions in
+    // the basic block in reverse post-order fashion.
+    for (auto I = BB->begin(), E = BB->end(); I != E; ++I) {
```
Is it actually necessary to try to sink the instruction as far into the block as possible as opposed to just placing it at the entry? It seems like anything that's going to block moving the instruction is also going to block improved pairing by the ARC optimizer, so it's not clear what value this has.

trentxintong replied:
It does not hurt to sink to as far as possible. also, this helps with retain coalescing on the LLVM level where we usually have less freedom to move retain and releases

trentxintong commented:
@rudkx I addressed your suggestions. Please tell me if there is anything else you think we should discuss. 

I am running 1 more test and benchmark before merging.

trentxintong commented:
@swift-ci Please benchmark

trentxintong commented:
@swift-ci Please test

trentxintong commented:
@swift-ci Please test

eeckstein commented:
Great work!
